File: lang/php/src/fun/vac/data_structures/LinkedStack.php
<?php
    namespace fun\vac\data_structures\LinkedStack;

    require_once "fun/vac/data_structures/interfaces/IStack.php";

    use fun\vac\data_structures\interfaces\IStack;

    class Node {

        public $data;
        public $next = NULL;

        public function __construct(int $element) {
            $this->data = $element;
        }
    }

    class LinkedStack implements IStack\IStack {

        private $head = NULL;
        private $tail = NULL;
        private $length = 0;

        public function __construct() {}

        public function size(): int {
            return $this->length;
        }

        public function isEmpty(): bool {
            return $this->length == 0;
        }

        public function peek(): int {
            if ($this->length == 0) {
                throw new RuntimeException("[PANIC - NoSuchElement]");
            }

            return $this->tail->data;
        }

        public function push(int $element): void {
            $node = new Node($element);

            if ($this->length == 0) {
                $this->head = $node;
            } else {
                $this->tail->next = $node;
            }

            $this->tail = $node;

            $this->length++;
        }

        public function pop(): void {
            if ($this->length == 0) {
                throw new RuntimeException("[PANIC - NoSuchElement]");
            }

            $target = $this->tail;

            if ($this->length == 1) {
                $this->head = NULL;
                $this->tail = NULL;
            } else {
                $cursor = $this->head;
                for ($i = 0, $bound = $this->length - 2; $i < $bound; $i++) {
                    $cursor = $cursor->next;
                }
                $cursor->next = NULL;
                $this->tail = $cursor;
            }

            $target->data = NULL;

            $this->length--;
        }
    }
?>

File: lang/java/src/fun/vac/data_structures/LinkedStack.java
package fun.vac.data_structures;

import fun.vac.data_structures.interfaces.IStack;

public class LinkedStack<E> implements IStack<E> {

    private static class Node<E> {

        public E data;
        public Node<E> next = null;

        public Node(E element) {
            data = element;
        }
    }

    private Node<E> head = null;
    private Node<E> tail = null;
    private int length = 0;

    public LinkedStack() {}

    @Override
    public int size() {
        return length;
    }

    @Override
    public boolean isEmpty() {
        return length == 0;
    }

    @Override
    public E peek() {
        if (length == 0) {
            throw new RuntimeException("[PANIC - NoSuchElement]");
        }

        return tail.data;
    }

    @Override
    @SuppressWarnings("unchecked")
    public void push(E element) {
        Node<E> node = new Node<>(element);

        if (length == 0) {
            head = node;
        } else {
            tail.next = node;
        }

        tail = node;

        length++;
    }

    @Override
    public void pop() {
        if (length == 0) {
            throw new RuntimeException("[PANIC - NoSuchElement]");
        }

        Node<E> target = tail;

        if (length == 1) {
            head = null;
            tail = null;
        } else {
            Node<E> cursor = head;
            for (int i = 0, bound = length - 2; i < bound; i++) {
                cursor = cursor.next;
            }
            cursor.next = null;
            tail = cursor;
        }

        target.data = null;

        length--;
    }
}
